processor/src/main/java/org/mapstruct/ap/internal/model/common/Parameter.java

```diff
@@ -100,4 +110,36 @@ public boolean equals(Object obj) {
         }
         return true;
     }
+
+    /**
+     * @param parameters the parameters to filter
+     * @return the parameters from the given list that are considered 'source parameters'
+     */
+    public static List<Parameter> getSourceParameters(List<Parameter> parameters) {
+        List<Parameter> sourceParameters = new ArrayList<Parameter>( parameters.size() );
+
+        for ( Parameter parameter : parameters ) {
+            if ( !parameter.isMappingTarget() && !parameter.isTargetType() && !parameter.isMappingContext() ) {
+                sourceParameters.add( parameter );
+            }
+        }
+
+        return sourceParameters;
+    }
+
+    /**
+     * @param parameters the parameters to filter
+     * @return the parameters from the given list that are marked as 'mapping context parameters'
+     */
+    public static List<Parameter> getContextParameters(List<Parameter> parameters) {
+        List<Parameter> contextParameters = new ArrayList<Parameter>( parameters.size() );
+
+        for ( Parameter parameter : parameters ) {
+            if ( parameter.isMappingContext() ) {
+                contextParameters.add( parameter );
+            }
+        }
+
+        return contextParameters;
```

Comment:
As an addition to @sjaakd, we can also profit from the 1.7 diamond generics in generated code when we detect Java 7. Maybe we need to create an issue where we can track this